### ISSUE TITLE
missing rule : space before the round brace for functions

### ISSUE THREAD

**Issue by remitbri** [eslint/eslint#1618]
Hi. I've tried to review each and every rule of JSHint, JSCS and ESLint to see if I could entirely "get rid of" the first two in favour of the latter one. As suggested by the owner of the @geteslint twitter account, I'm reporting the JSCS rules I've found actually not covered by ESLint. My apologies if it's a duplicate.

 - "disallowSpacesInAnonymousFunctionExpression"
 - "requireSpacesInAnonymousFunctionExpression"
 - "disallowSpacesInNamedFunctionExpression"
 - "requireSpacesInNamedFunctionExpression"
 - "disallowSpacesInNamedFunctionExpression"
 - "requireSpacesInNamedFunctionExpression"

The space before the opening curly brace is covered by "space-before-block".
What is not covered is the space before the round brace, i.e being able to say this is invalid

```js
var Foobar = {
    foo: function (){}
}
array.map(function () {})
var x = function () {}
var x = function a (){}
function a () {}
```

only because, say, the rule would require a space between the function name (or `function` if it is anonymous) and the opening round brace. Or the contrary.
And perhaps being able to differentiate between named and anonymous functions.

**Comment by nzakas** [comment]
By "round brace" are you talking about the parentheses?

Can you see if this rule does what you're looking for? http://eslint.org/docs/rules/space-after-keywords.html

**Comment by lo1tuma** [comment]
This should be covered by [space-after-keywords](http://eslint.org/docs/rules/space-after-keywords.html) with the `checkFunctionKeyword` option enabled and   by [space-after-function-name](http://eslint.org/docs/rules/space-after-function-name.html).

**Comment by remitbri** [comment]
Oops! I really missed `space-after-function-name` indeed. Thanks.

Hmm, what about anonymous functions? if space is required, there's `space-after-keywords`, but then what if I need a space after `if`, `else`, etc, but not in `function() {}` ?

**Comment by lo1tuma** [comment]
You are right, this case is not possible with those two rules. I consider this as an inconsistent coding style like enforcing a space after `if` and disallow it after `while`.

**Comment by remitbri** [comment]
I see. Well, food for thoughts, thanks! I guess all bases are covered.

**Comment by nzakas** [comment]
I think we should find a way to cover that case. Anonymous function styling is a special case in most situations.

**Comment by doberkofler** [comment]
+1

**Comment by message** [comment]
+1

**Comment by message** [comment]
@nzakas let's discuss this? Should this be separate rule or `space-after-function-name`?

**Comment by nzakas** [comment]
I think it should be part of `space-after-function-name`, allowing you specify what you want for anonymous functions. That wouldn't be a breaking change because it would be off by default.

**Comment by message** [comment]
@nzakas Working on this. Will deliver in an hour.

**Comment by lo1tuma** [comment]
@nzakas can you explain why you think this should be part of `space-after-function-name`? How is the space after the `function` keyword related to the function name, which does not exist for anonymous functions.

**Comment by message** [comment]
`[1, "always", {"anonymous":true}]` makes sense to me

**Comment by lo1tuma** [comment]
@message Sorry, I don’t get it. You try to enforce a space after a function name for an anonymous function which doesn’t have a function name.

**Comment by lo1tuma** [comment]
Another reason not to do this in `space-after-function-name` is if you want to enforce the following patterns:

```js
function foo () {}
var bar = function() {}
```

**Comment by nzakas** [comment]
@lotuma The goal here is to have one place where someone can go to configure the space before the parentheses. @message's approach is roughly what I had in mind, and would result in your code passing if "anonymous" was `false`.

As already mentioned in this thread, `function` is a special case that doesn't really fit with the other keywords, so we have a pattern that is uncovered.

Do you have a suggestion for something that makes more sense to you?

**Comment by message** [comment]
@nzakas can you please review my commit?

**Comment by doberkofler** [comment]
@lo1tuma and what about the option of enforcing the pattern used by Douglas Crockford in JSLint ?

```
function foo() {}
var bar = function () {};
```

**Comment by lo1tuma** [comment]
@nzakas How does it pass my example code above? In one case it enforce a space after the function name, in the other case it disallows a space before the parentheses.

I agree that we should only have one place to configure this, but `space-after-function-name` would be a confusing name. ESLint already has a lot of confusing rule and option names, we should not make it worse.
So instead of `space-after-function-name` we want a rule `space-before-function-parentheses` with the following options:

* `"always"` enforces a space before function parentheses for named functions and anonymous functions
* `"never"` disallows a space before function parentheses for named functions and anonymous functions
* `{"anonymous": "always", "named": "never"}` enforces a space before the parentheses of anonymous functions but disallows a space before the parentheses of named functions
* `{"anonymous": "never", "named": "always"}` disallows a space before the parentheses of anonymous functions but enforces a space before the parentheses of named functions

Additionally we can remove the `checkFunctionKeyword` from the `space-after-keyword` rule.

@doberkofler This is already possible in ESLint.

**Comment by message** [comment]
@lo1tuma yeap, that will be cleaner

**Comment by nzakas** [comment]
@lo1tuma the `[1, "always", {"anonymous": false}]` would say " always have a space after the name unless it's an anonymous function, then there should be on space."

Anyway, I like the idea of `space-before-function-parentheses` better. Let's do that.

**Comment by message** [comment]
@lo1tuma can you implement your idea?

**Comment by lo1tuma** [comment]
@nzakas Thanks, no it makes sense, anyway I still think `space-before-function-parentheses` is better because it also remove the `checkFunctionKeyword` from  `space-after-keyword`.

Working on it.
